File: tests/fixtures/doorbell/before/zircon/kernel/lib/syscalls/doorbell.cc
// Copyright 2026 The Fuchsia Authors
//
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT

// Synthetic syscalls used to test babeldiff. They are not part of Zircon.

#include <lib/syscalls/forward.h>
#include <trace.h>

#include <object/doorbell_dispatcher.h>
#include <object/process_dispatcher.h>

#define LOCAL_TRACE 0

// zx_status_t zx_doorbell_ring
zx_status_t sys_doorbell_ring(zx_handle_t handle, uint32_t tone,
                              user_out_ptr<uint64_t> out_count) {
  LTRACEF("handle %x tone %u\n", handle, tone);

  zx_status_t status;
  auto up = ProcessDispatcher::GetCurrent();
  fbl::RefPtr<DoorbellDispatcher> doorbell;
  status = up->handle_table().GetDispatcherWithRights(*up, handle, ZX_RIGHT_SIGNAL, &doorbell);
  if (status != ZX_OK) {
    return status;
  }

  status = doorbell->Ring(tone);
  if (status == ZX_OK && out_count) {
    status = out_count.copy_to_user(doorbell->RingCount());
  }

  return status;
}

File: tests/fixtures/doorbell/after/zircon/kernel/lib/syscalls/doorbell.rs
// Copyright 2026 The Fuchsia Authors
//
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT

// Synthetic syscalls used to test babeldiff. They are not part of Zircon.

use crate::object::dispatcher::Dispatcher;
use crate::object::doorbell_dispatcher::DoorbellDispatcher;
use crate::user_copy::UserOutPtr;
use zx_status::Status;
use zx_types::{HandleValue, ZX_RIGHT_SIGNAL};

const LOCAL_TRACE: bool = false;

pub fn sys_doorbell_ring(
    handle: HandleValue,
    tone: u32,
    out_count: UserOutPtr<u64>,
) -> Result<(), Status> {
    ltracef!("handle {:?} tone {}\n", handle, tone);

    let doorbell = Dispatcher::get_with_rights::<DoorbellDispatcher>(handle, ZX_RIGHT_SIGNAL)?;
    doorbell.ring(tone)?;
    if !out_count.is_null() {
        out_count.write(doorbell.ring_count())?;
    }
    Ok(())
}
